File: REFERENSI/cassette_detector.cpp
#include "cassette_detector.h"
#include <iostream>
#include <sstream>
#include <regex>
#include <iomanip>
#include <algorithm>
#include <ctime>

CassetteDetector::CassetteDetector(const std::string& data) : xmlData(data) {
    dispensedValues.resize(MAX_CASSETTES, 0);
    rejectedValues.resize(MAX_CASSETTES, 0);
    remainingValues.resize(MAX_CASSETTES, 0);
    denominations.resize(MAX_CASSETTES, 0);
    initialValues.resize(MAX_CASSETTES, 0);
    denominationsView.resize(MAX_CASSETTES, "");
    cassetteTypes.resize(MAX_CASSETTES, 0);
}

std::string CassetteDetector::formatCurrency(int value, const std::string& currency) {
    if (value == 0) return currency + " 0";

    std::ostringstream oss;
    oss << currency + " ";

    std::string numStr = std::to_string(value);
    std::string formatted;
    int count = 0;

    for (int i = static_cast<int>(numStr.length()) - 1; i >= 0; i--) {
        if (count > 0 && count % 3 == 0) {
            formatted = "." + formatted;
        }
        formatted = numStr[i] + formatted;
        count++;
    }

    oss << formatted;
    return oss.str();
}

std::string CassetteDetector::padRight(const std::string& str, int width) {
    if (static_cast<int>(str.length()) >= width) return str;
    return str + std::string(width - str.length(), ' ');
}

std::string CassetteDetector::padCenter(const std::string& str, int width) {
    if (static_cast<int>(str.length()) >= width) return str;
    int padding = width - static_cast<int>(str.length());
    int leftPad = padding / 2;
    int rightPad = padding - leftPad;
    return std::string(leftPad, ' ') + str + std::string(rightPad, ' ');
}

void CassetteDetector::parseXMLData() {
    std::regex timestampPattern(R"(^\d{2}\.\d{2}\.\d{4} \d{2}:\d{2}:\d{2}\.\d{3} \[TID:0x[0-9A-Fa-f]+\] \[XML:\d+\] )");
    std::string cleanedXML = std::regex_replace(xmlData, timestampPattern, "");

    std::regex attrPattern(R"((\w+)=([\w|]+))");
    cleanedXML = std::regex_replace(cleanedXML, attrPattern, "$1='$2'");

    parseXMLElements(cleanedXML);
}

void CassetteDetector::parseXMLElements(const std::string& xml) {
    std::regex cuPattern(R"(<cu\s+([^>]+)>)");
    std::sregex_iterator iter(xml.begin(), xml.end(), cuPattern);
    std::sregex_iterator end;

    int dcIndex = 1;

    for (; iter != end; ++iter) {
        std::string attributes = iter->str(1);
        std::map<std::string, std::string> attrs = parseAttributes(attributes);

        if (attrs.find("n") == attrs.end()) continue;

        int vIndex = std::stoi(attrs["n"]) - 1;
        if (vIndex < 0 || vIndex >= MAX_CASSETTES) continue;

        int vaIndex = std::stoi(attrs["v"]);
        int typIndex = std::stoi(attrs["typ"]);
        int denominationValue = std::stoi(attrs["v"]);
        std::string dcView = attrs["dc"];

        int parsedDcIndex = dcIndex;
        if (attrs["dc"] != "0" && !attrs["dc"].empty() && std::isdigit(attrs["dc"][0])) {
            parsedDcIndex = std::stoi(attrs["dc"]);
        }
        else {
            parsedDcIndex = dcIndex++;
        }

        if (cashTotalLabels.find(parsedDcIndex) == cashTotalLabels.end()) {
            if (typIndex == 2) {
                cashTotalLabels[parsedDcIndex] = "REJECT_" + dcView;
            }
            else if (typIndex == 3) {
                cashTotalLabels[parsedDcIndex] = "CASSETTE_" + dcView;
            }
            else if (typIndex == 6) {
                cashTotalLabels[parsedDcIndex] = "RETRACT_" + dcView;
            }
        }

        if (denominationValues.find(vaIndex) == denominationValues.end()) {
            denominationValues[vaIndex] = denominationValue;
        }

        initialValues[vIndex] = attrs.count("ic") ? std::stoi(attrs["ic"]) : 0;
        dispensedValues[vIndex] = attrs.count("ds") ? std::stoi(attrs["ds"]) : 0;
        rejectedValues[vIndex] = attrs.count("rj") ? std::stoi(attrs["rj"]) : 0;
        remainingValues[vIndex] = attrs.count("c") ? std::stoi(attrs["c"]) : 0;
        denominations[vIndex] = denominationValue / 100;
        denominationsView[vIndex] = formatCurrency((denominationValue / 100),"");
        cassetteTypes[vIndex] = typIndex;
    }
}

std::map<std::string, std::string> CassetteDetector::parseAttributes(const std::string& attrs) {
    std::map<std::string, std::string> result;
    std::regex attrRegex(R"((\w+)=['"](.*?)['"])");
    std::sregex_iterator iter(attrs.begin(), attrs.end(), attrRegex);
    std::sregex_iterator end;

    for (; iter != end; ++iter) {
        result[iter->str(1)] = iter->str(2);
    }

    return result;
}

std::string CassetteDetector::detectCassette() {
    parseXMLData();

    std::ostringstream sb;

    sb << "================================================================================\n";
    sb << "                          CASH DISPENSER STATUS REPORT\n";
    sb << "================================================================================\n";

    std::vector<int> activeCassettes;
    for (int i = 0; i < MAX_CASSETTES; i++) {
        if (initialValues[i] > 0 || dispensedValues[i] > 0 || rejectedValues[i] > 0 || remainingValues[i] > 0) {
            activeCassettes.push_back(i);
        }
    }

    std::vector<int> financialCassettes;
    for (int i : activeCassettes) {
        if (cassetteTypes[i] != 2 && cassetteTypes[i] != 6) {
            financialCassettes.push_back(i);
        }
    }

    if (activeCassettes.empty()) {
        sb << "No active cassettes detected.\n";
        sb << "================================================================================\n";
        return sb.str();
    }

    sb << "\nCASETTE SUMMARY:\n";
    sb << "--------------------------------------------------------------------------------\n";
    sb << "CASSETTE        DENOMINATION    INITIAL    DISPENSED   REJECTED   REMAINING\n";
    sb << "--------------------------------------------------------------------------------\n";

    for (int cassetteIndex : activeCassettes) {
        auto labelIt = std::find_if(cashTotalLabels.begin(), cashTotalLabels.end(),
            [cassetteIndex](const std::pair<int, std::string>& p) {
                return p.first == cassetteIndex + 1;
            });

        std::string cassetteLabel = (labelIt != cashTotalLabels.end()) ?
            labelIt->second : "CASSETTE_" + std::to_string(cassetteIndex + 1);

        sb << std::left << std::setw(16) << cassetteLabel.substr(0, 16)
            << std::setw(16) << denominationsView[cassetteIndex]
            << std::setw(11) << initialValues[cassetteIndex]
                << std::setw(12) << dispensedValues[cassetteIndex]
                    << std::setw(11) << rejectedValues[cassetteIndex]
                        << std::setw(10) << remainingValues[cassetteIndex] << "\n";
    }

    sb << "--------------------------------------------------------------------------------\n";

    int totalInitial = 0;
    int totalDispensed = 0;
    int totalRejected = 0;
    int totalRemaining = 0;
    bool hasDiscrepancy = false;

    sb << "\nSUMMARY PER DISPENSER CASSETTE:\n";
    sb << "--------------------------------------------------------------------------------\n";

    if (financialCassettes.empty()) {
        sb << "No dispenser cassettes found (only reject/retract cassettes detected).\n";
    }
    else {
        for (int cassetteIndex : financialCassettes) {
            auto labelIt = std::find_if(cashTotalLabels.begin(), cashTotalLabels.end(),
                [cassetteIndex](const std::pair<int, std::string>& p) {
                    return p.first == cassetteIndex + 1;
                });

            std::string cassetteLabel = (labelIt != cashTotalLabels.end()) ?
                labelIt->second : "CASSETTE_" + std::to_string(cassetteIndex + 1);

            int denomination = denominations[cassetteIndex];
            int initial = initialValues[cassetteIndex] * denomination;
            int dispensed = dispensedValues[cassetteIndex] * denomination;
            int rejected = rejectedValues[cassetteIndex] * denomination;
            int remaining = remainingValues[cassetteIndex] * denomination;
            int total = remaining + dispensed + rejected;

            std::string typeInfo = "";
            if (cassetteTypes[cassetteIndex] == 3) {
                typeInfo = " (DISPENSER)";
            }

            sb << cassetteLabel << typeInfo << ":\n";
            sb << "  Initial Amount   : " << formatCurrency(initial, "Rp") << "\n";
            sb << "  Dispensed Amount : " << formatCurrency(dispensed, "Rp") << "\n";
            sb << "  Rejected Amount  : " << formatCurrency(rejected, "Rp") << "\n";
            sb << "  Remaining Amount : " << formatCurrency(remaining, "Rp") << "\n";
            sb << "  Total Calculated : " << formatCurrency(total, "Rp") << "\n";

            if (initial > 0 && total != initial) {
                hasDiscrepancy = true;
                sb << "  \n*** SUSPECT JAMMED DETECTED ***\n";
                sb << "  TOTAL COUNTER NOT MATCH " << formatCurrency(total, "Rp")
                    << " WITH INITIAL COUNT " << formatCurrency(initial, "Rp")
                    << " ON " << cassetteLabel << "\n";
                sb << "  DIFFERENCE: " << formatCurrency(abs(total - initial), "Rp") << "\n";
            }

            totalInitial += initial;
            totalDispensed += dispensed;
            totalRejected += rejected;
            totalRemaining += remaining;
        }
    }

    std::vector<int> excludedCassettes;
    for (int i : activeCassettes) {
        if (cassetteTypes[i] == 2 || cassetteTypes[i] == 6) {
            excludedCassettes.push_back(i);
        }
    }

    if (denominationValues.size() > 1 && !financialCassettes.empty()) {
        sb << "\nSUMMARY BY DENOMINATION:\n";
        sb << "--------------------------------------------------------------------------------\n";

        for (const auto& kvp : denominationValues) {
            int denomination = kvp.second / 100;
            if (denomination == 0) continue;

            int denomInitial = 0, denomDispensed = 0, denomRejected = 0, denomRemaining = 0;

            for (int i = 0; i < MAX_CASSETTES; i++) {
                if (denominations[i] == denomination && cassetteTypes[i] != 2 && cassetteTypes[i] != 6) {
                    denomInitial += initialValues[i] * denomination;
                    denomDispensed += dispensedValues[i] * denomination;
                    denomRejected += rejectedValues[i] * denomination;
                    denomRemaining += remainingValues[i] * denomination;
                }
            }

            if (denomInitial > 0 || denomDispensed > 0 || denomRejected > 0 || denomRemaining > 0) {
                sb << "Denomination" << formatCurrency(denomination, "") << " :\n";
                sb << "  Initial   : " << formatCurrency(denomInitial, "Rp") << "\n";
                sb << "  Dispensed : " << formatCurrency(denomDispensed, "Rp") << "\n";
                sb << "  Rejected  : " << formatCurrency(denomRejected, "Rp") << "\n";
                sb << "  Remaining : " << formatCurrency(denomRemaining, "Rp") << "\n";

                int denomTotal = denomRemaining + denomDispensed + denomRejected;
                if (denomInitial > 0 && denomTotal != denomInitial) {
                    hasDiscrepancy = true;
                    sb << "  \n*** SUSPECT JAMMED DETECTED ***\n";
                    sb << "  TOTAL COUNTER NOT MATCH " << formatCurrency(denomTotal, "Rp")
                        << " WITH INITIAL COUNT " << formatCurrency(denomInitial, "Rp")
                        << " ON DENOMINATION " << formatCurrency(denomination, "") << "\n";
                }
            }
        }
    }

    if (!financialCassettes.empty()) {
        sb << "\nOVERALL CASH DISPENSER SUMMARY:\n";
        sb << "--------------------------------------------------------------------------------\n";
        sb << "Total Initial Amount    : " << formatCurrency(totalInitial, "Rp") << "\n";
        sb << "Total Dispensed Amount  : " << formatCurrency(totalDispensed, "Rp") << "\n";
        sb << "Total Rejected Amount   : " << formatCurrency(totalRejected, "Rp") << "\n";
        sb << "Total Remaining Amount  : " << formatCurrency(totalRemaining, "Rp") << "\n";
        sb << "Total in Dispenser      : " << formatCurrency(totalDispensed + totalRejected + totalRemaining, "Rp") << "\n";

        int overallTotal = totalDispensed + totalRejected + totalRemaining;
        if (totalInitial > 0 && overallTotal != totalInitial) {
            hasDiscrepancy = true;
            sb << "\n*** OVERALL DISCREPANCY DETECTED ***\n";
            sb << "Expected Total: " << formatCurrency(totalInitial, "Rp") << "\n";
            sb << "Actual Total:   " << formatCurrency(overallTotal, "Rp") << "\n";
            sb << "Difference:     " << formatCurrency(abs(overallTotal - totalInitial), "Rp") << "\n";
        }
    }

    if (!financialCassettes.empty()) {
        if (hasDiscrepancy) {
            sb << "\n*** ATTENTION: CASSETTE DISCREPANCIES DETECTED ***\n";
            sb << "Please check the dispenser for possible jamming or mechanical issues.\n";
        }
    }

    sb << "\n";
    sb << "================================================================================";

    return sb.str();
}

bool CassetteProcessor::isCassetteData(const std::string& line) {
    return line.find("[XML:15]") != std::string::npos &&
        line.find("<cunits") != std::string::npos;
}

std::string CassetteProcessor::processCassetteXML(const std::string& line) {
    if (!isCassetteData(line)) {
        return "";
    }
    CassetteDetector detector(line);
    std::string result = detector.detectCassette();

    std::regex lfRegex("\n");
    result = std::regex_replace(result, lfRegex, "\r\n");

    return result;
}

File: JalinTools/Services/CassetteDetector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace JalinTools.Services
{
    /// <summary>
    /// Detects and formats cassette information from XML data
    /// Based on C++ cassette_detector implementation
    /// </summary>
    public class CassetteDetector
    {
        private const int MAX_CASSETTES = 5;
        
        private readonly string xmlData;
        private readonly int[] dispensedValues = new int[MAX_CASSETTES];
        private readonly int[] rejectedValues = new int[MAX_CASSETTES];
        private readonly int[] remainingValues = new int[MAX_CASSETTES];
        private readonly int[] denominations = new int[MAX_CASSETTES];
        private readonly int[] initialValues = new int[MAX_CASSETTES];
        private readonly string[] denominationsView = new string[MAX_CASSETTES];
        private readonly int[] cassetteTypes = new int[MAX_CASSETTES];
        private readonly Dictionary<int, string> cashTotalLabels = new Dictionary<int, string>();
        private readonly Dictionary<int, int> denominationValues = new Dictionary<int, int>();

        public CassetteDetector(string data)
        {
            xmlData = data;
            for (int i = 0; i < MAX_CASSETTES; i++)
            {
                denominationsView[i] = "";
            }
        }

        public string DetectCassette()
        {
            ParseXMLData();
            
            var sb = new StringBuilder();
            
            sb.AppendLine("================================================================================");
            sb.AppendLine("                     LAPORAN STATUS CASH DISPENSER");
            sb.AppendLine("================================================================================");
            
            // Find active cassettes
            var activeCassettes = new List<int>();
            for (int i = 0; i < MAX_CASSETTES; i++)
            {
                if (initialValues[i] > 0 || dispensedValues[i] > 0 || rejectedValues[i] > 0 || remainingValues[i] > 0)
                {
                    activeCassettes.Add(i);
                }
            }
            
            // Find financial cassettes (not reject/retract)
            var financialCassettes = activeCassettes
                .Where(i => cassetteTypes[i] != 2 && cassetteTypes[i] != 6)
                .ToList();
            
            if (activeCassettes.Count == 0)
            {
                sb.AppendLine("Tidak ada kaset aktif terdeteksi.");
                sb.AppendLine("================================================================================");
                return sb.ToString();
            }
            
            // CASSETTE SUMMARY TABLE
            sb.AppendLine();
            sb.AppendLine("RINGKASAN KASET:");
            sb.AppendLine("--------------------------------------------------------------------------------");
            sb.AppendLine("KASET           DENOMINASI      AWAL       KELUAR      REJECT      SISA");
            sb.AppendLine("--------------------------------------------------------------------------------");
            
            foreach (var cassetteIndex in activeCassettes)
            {
                var label = cashTotalLabels.ContainsKey(cassetteIndex + 1)
                    ? cashTotalLabels[cassetteIndex + 1]
                    : $"CASSETTE_{cassetteIndex + 1}";
                
                sb.AppendLine(
                    $"{label.PadRight(16)}" +
                    $"{denominationsView[cassetteIndex].PadRight(16)}" +
                    $"{initialValues[cassetteIndex].ToString().PadRight(11)}" +
                    $"{dispensedValues[cassetteIndex].ToString().PadRight(12)}" +
                    $"{rejectedValues[cassetteIndex].ToString().PadRight(12)}" +
                    $"{remainingValues[cassetteIndex]}");
            }
            
            sb.AppendLine("--------------------------------------------------------------------------------");
            
            // SUMMARY PER DISPENSER CASSETTE
            int totalInitial = 0;
            int totalDispensed = 0;
            int totalRejected = 0;
            int totalRemaining = 0;
            bool hasDiscrepancy = false;
            
            sb.AppendLine();
            sb.AppendLine("RINGKASAN PER KASET DISPENSER:");
            sb.AppendLine("--------------------------------------------------------------------------------");
            
            if (financialCassettes.Count == 0)
            {
                sb.AppendLine("Tidak ada kaset dispenser (hanya kaset reject/retract terdeteksi).");
            }
            else
            {
                foreach (var cassetteIndex in financialCassettes)
                {
                    var label = cashTotalLabels.ContainsKey(cassetteIndex + 1)
                        ? cashTotalLabels[cassetteIndex + 1]
                        : $"CASSETTE_{cassetteIndex + 1}";
                    
                    int denomination = denominations[cassetteIndex];
                    int initial = initialValues[cassetteIndex] * denomination;
                    int dispensed = dispensedValues[cassetteIndex] * denomination;
                    int rejected = rejectedValues[cassetteIndex] * denomination;
                    int remaining = remainingValues[cassetteIndex] * denomination;
                    int total = remaining + dispensed + rejected;
                    
                    string typeInfo = cassetteTypes[cassetteIndex] == 3 ? " (DISPENSER)" : "";
                    
                    sb.AppendLine($"{label}{typeInfo}:");
                    sb.AppendLine($"  Jumlah Awal     : {FormatCurrency(initial, "Rp")}");
                    sb.AppendLine($"  Jumlah Keluar   : {FormatCurrency(dispensed, "Rp")}");
                    sb.AppendLine($"  Jumlah Reject   : {FormatCurrency(rejected, "Rp")}");
                    sb.AppendLine($"  Jumlah Sisa     : {FormatCurrency(remaining, "Rp")}");
                    sb.AppendLine($"  Total Terhitung : {FormatCurrency(total, "Rp")}");
                    
                    if (initial > 0 && total != initial)
                    {
                        hasDiscrepancy = true;
                        sb.AppendLine();
                        sb.AppendLine("  *** DICURIGAI ADA JAMMED ***");
                        sb.AppendLine($"  TOTAL COUNTER TIDAK COCOK {FormatCurrency(total, "Rp")}" +
                            $" DENGAN JUMLAH AWAL {FormatCurrency(initial, "Rp")}" +
                            $" PADA {label}");
                        sb.AppendLine($"  SELISIH: {FormatCurrency(Math.Abs(total - initial), "Rp")}");
                    }
                    
                    totalInitial += initial;
                    totalDispensed += dispensed;
                    totalRejected += rejected;
                    totalRemaining += remaining;
                }
            }
            
            // SUMMARY BY DENOMINATION
            if (denominationValues.Count > 1 && financialCassettes.Count > 0)
            {
                sb.AppendLine();
                sb.AppendLine("RINGKASAN PER DENOMINASI:");
                sb.AppendLine("--------------------------------------------------------------------------------");
                
                foreach (var kvp in denominationValues.OrderByDescending(x => x.Value))
                {
                    int denomination = kvp.Value / 100;
                    if (denomination == 0) continue;
                    
                    int denomInitial = 0, denomDispensed = 0, denomRejected = 0, denomRemaining = 0;
                    
                    for (int i = 0; i < MAX_CASSETTES; i++)
                    {
                        if (denominations[i] == denomination && cassetteTypes[i] != 2 && cassetteTypes[i] != 6)
                        {
                            denomInitial += initialValues[i] * denomination;
                            denomDispensed += dispensedValues[i] * denomination;
                            denomRejected += rejectedValues[i] * denomination;
                            denomRemaining += remainingValues[i] * denomination;
                        }
                    }
                    
                    if (denomInitial > 0 || denomDispensed > 0 || denomRejected > 0 || denomRemaining > 0)
                    {
                        sb.AppendLine($"Denominasi {FormatCurrency(denomination, "")} :");
                        sb.AppendLine($"  Awal   : {FormatCurrency(denomInitial, "Rp")}");
                        sb.AppendLine($"  Keluar : {FormatCurrency(denomDispensed, "Rp")}");
                        sb.AppendLine($"  Reject : {FormatCurrency(denomRejected, "Rp")}");
                        sb.AppendLine($"  Sisa   : {FormatCurrency(denomRemaining, "Rp")}");
                        
                        int denomTotal = denomRemaining + denomDispensed + denomRejected;
                        if (denomInitial > 0 && denomTotal != denomInitial)
                        {
                            hasDiscrepancy = true;
                            sb.AppendLine();
                            sb.AppendLine("  *** DICURIGAI ADA JAMMED ***");
                            sb.AppendLine($"  TOTAL COUNTER TIDAK COCOK {FormatCurrency(denomTotal, "Rp")}" +
                                $" DENGAN JUMLAH AWAL {FormatCurrency(denomInitial, "Rp")}" +
                                $" PADA DENOMINASI {FormatCurrency(denomination, "")}");
                        }
                    }
                }
            }
            
            // OVERALL SUMMARY
            if (financialCassettes.Count > 0)
            {
                sb.AppendLine();
                sb.AppendLine("RINGKASAN TOTAL CASH DISPENSER:");
                sb.AppendLine("--------------------------------------------------------------------------------");
                sb.AppendLine($"Total Jumlah Awal       : {FormatCurrency(totalInitial, "Rp")}");
                sb.AppendLine($"Total Jumlah Keluar     : {FormatCurrency(totalDispensed, "Rp")}");
                sb.AppendLine($"Total Jumlah Reject     : {FormatCurrency(totalRejected, "Rp")}");
                sb.AppendLine($"Total Jumlah Sisa       : {FormatCurrency(totalRemaining, "Rp")}");
                sb.AppendLine($"Total dalam Dispenser   : {FormatCurrency(totalDispensed + totalRejected + totalRemaining, "Rp")}");
                
                int overallTotal = totalDispensed + totalRejected + totalRemaining;
                if (totalInitial > 0 && overallTotal != totalInitial)
                {
                    hasDiscrepancy = true;
                    sb.AppendLine();
                    sb.AppendLine("*** KETIDAKCOCOKAN TOTAL TERDETEKSI ***");
                    sb.AppendLine($"Total Yang Diharapkan: {FormatCurrency(totalInitial, "Rp")}");
                    sb.AppendLine($"Total Aktual:          {FormatCurrency(overallTotal, "Rp")}");
                    sb.AppendLine($"Selisih:               {FormatCurrency(Math.Abs(overallTotal - totalInitial), "Rp")}");
                }
            }
            
            if (financialCassettes.Count > 0 && hasDiscrepancy)
            {
                sb.AppendLine();
                sb.AppendLine("*** PERHATIAN: KETIDAKCOCOKAN KASET TERDETEKSI ***");
                sb.AppendLine("Silakan periksa dispenser untuk kemungkinan jamming atau masalah mekanis.");
            }
            
            sb.AppendLine();
            sb.Append("================================================================================");
            
            return sb.ToString();
        }

        private void ParseXMLData()
        {
            // Clean timestamp and tags
            var timestampPattern = new Regex(@"^\d{2}\.\d{2}\.\d{4} \d{2}:\d{2}:\d{2}\.\d{3} \[TID:0x[0-9A-Fa-f]+\] \[XML:\d+\] ");
            string cleanedXML = timestampPattern.Replace(xmlData, "");
            
            // Add quotes to attributes
            var attrPattern = new Regex(@"(\w+)=([\w|]+)");
            cleanedXML = attrPattern.Replace(cleanedXML, "$1='$2'");
            
            ParseXMLElements(cleanedXML);
        }

        private void ParseXMLElements(string xml)
        {
            var cuPattern = new Regex(@"<cu\s+([^>]+)>");
            var matches = cuPattern.Matches(xml);
            
            int dcIndex = 1;
            
            foreach (Match match in matches)
            {
                var attributes = match.Groups[1].Value;
                var attrs = ParseAttributes(attributes);
                
                if (!attrs.ContainsKey("n")) continue;
                
                int vIndex = int.Parse(attrs["n"]) - 1;
                if (vIndex < 0 || vIndex >= MAX_CASSETTES) continue;
                
                int typIndex = int.Parse(attrs["typ"]);
                int denominationValue = int.Parse(attrs["v"]);
                string dcView = attrs.ContainsKey("dc") ? attrs["dc"] : "0";
                
                int parsedDcIndex = dcIndex;
                if (dcView != "0" && !string.IsNullOrEmpty(dcView) && char.IsDigit(dcView[0]))
                {
                    parsedDcIndex = int.Parse(dcView);
                }
                else
                {
                    parsedDcIndex = dcIndex++;
                }
                
                if (!cashTotalLabels.ContainsKey(parsedDcIndex))
                {
                    if (typIndex == 2)
                        cashTotalLabels[parsedDcIndex] = "REJECT_" + dcView;
                    else if (typIndex == 3)
                        cashTotalLabels[parsedDcIndex] = "CASSETTE_" + dcView;
                    else if (typIndex == 6)
                        cashTotalLabels[parsedDcIndex] = "RETRACT_" + dcView;
                }
                
                if (!denominationValues.ContainsKey(vIndex))
                {
                    denominationValues[vIndex] = denominationValue;
                }
                
                initialValues[vIndex] = attrs.ContainsKey("ic") ? int.Parse(attrs["ic"]) : 0;
                dispensedValues[vIndex] = attrs.ContainsKey("ds") ? int.Parse(attrs["ds"]) : 0;
                rejectedValues[vIndex] = attrs.ContainsKey("rj") ? int.Parse(attrs["rj"]) : 0;
                remainingValues[vIndex] = attrs.ContainsKey("c") ? int.Parse(attrs["c"]) : 0;
                denominations[vIndex] = denominationValue / 100;
                denominationsView[vIndex] = FormatCurrency(denominationValue / 100, "");
                cassetteTypes[vIndex] = typIndex;
            }
        }

        private Dictionary<string, string> ParseAttributes(string attrs)
        {
            var result = new Dictionary<string, string>();
            var attrRegex = new Regex(@"(\w+)=['""](.+?)['""]");
            var matches = attrRegex.Matches(attrs);
            
            foreach (Match match in matches)
            {
                result[match.Groups[1].Value] = match.Groups[2].Value;
            }
            
            return result;
        }

        private string FormatCurrency(int value, string currency)
        {
            if (value == 0) return currency + (string.IsNullOrEmpty(currency) ? "" : " ") + "0";
            
            string numStr = value.ToString();
            string formatted = "";
            int count = 0;
            
            for (int i = numStr.Length - 1; i >= 0; i--)
            {
                if (count > 0 && count % 3 == 0)
                {
                    formatted = "." + formatted;
                }
                formatted = numStr[i] + formatted;
                count++;
            }
            
            return string.IsNullOrEmpty(currency) ? formatted : currency + " " + formatted;
        }

        public static bool IsCassetteData(string line)
        {
            return line.Contains("[XML:15]") && line.Contains("<cunits");
        }

        public static string ProcessCassetteXML(string line)
        {
            if (!IsCassetteData(line)) return string.Empty;
            
            var detector = new CassetteDetector(line);
            return detector.DetectCassette();
        }
    }
}
